### PR TITLE
Bowling Challenge

### PR DESCRIPTION
**Pull request by Caroline-Amanquah**

### CODE REVIEW

File: bowling.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+          extraBalls = 2 * this.getRandomScore();
+          this.scores[this.frames - 1] += extraBalls;
+        } else if (this.scores[this.frames - 1] > this.totalPins) {
+          extraBalls = this.getRandomScore();
+          this.scores[this.frames - 1] += extraBalls;
+        }
+      }
+      
+      
+    getRandomScore(limit = this.totalPins) {
```
Comment (charlie-lam) [inline bowling.js:55]: Nice separation of concerns.

File: bowling.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      return Math.floor(Math.random() * limit);
+    }
+
```
Comment (charlie-lam) [inline bowling.js:58]: Also nice random number generator function.

File: bowling.test.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+const BowlingGame  = require('./bowling');
+
+
+describe("playFrame", () => {
+    it("scores a strike", () => {
+        const game = new BowlingGame(10, 2, 10);
+        game.getRandomScore = jest.fn(() => 10);
```
Comment (charlie-lam) [inline bowling.test.js:7]: Use of jest mocks is great.

**Review by charlie-lam — COMMENTED** [review]
Maybe not bowling but nice code.

File: bowling.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
+const frames = 10;
+const rollsPerFrame = 2;
+const totalPins = 10;
+let ball1 = 0;
+let ball2 = 0;
```
Comment (z-rasool) [inline bowling.js:5]: Instead of variables, could Balls be a separate class?

File: bowling.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      return Math.floor(Math.random() * limit);
+    }
+  
+    calculateTotalScore() {
+      return this.scores.reduce((a, b) => a + b, 0);
+    }
+  }
+  
+  const game = new BowlingGame(10, 2, 10);
+  game.play();
```
Comment (z-rasool) [inline bowling.js:65]: Did you intend to hardcode the play in this class? You might want to remove this so it doesn't interfere with any tests!

File: bowling.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        this.playFrame();
+      }
+      this.playExtraBalls();
+      console.log("Total Score: ", this.calculateTotalScore());
+    }
+  
+    playFrame() {
+      let ball1 = this.getRandomScore();
+      if (ball1 === this.totalPins) {
+        this.scores[this.currentFrame] = this.totalPins + this.getRandomScore() + this.getRandomScore();
```
Comment (z-rasool) [inline bowling.js:28]: Could you refactor the repetition of `this.getRandomScore()` using `*2`?

File: bowling.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      let ball2 = this.getRandomScore(this.totalPins - ball1);
+      let frameScore = 0;
+      if (ball1 + ball2 === this.totalPins) {
+        frameScore = this.totalPins + this.getRandomScore();
+      } else {
+        frameScore = ball1 + ball2;
+      }
+      this.scores[this.currentFrame] = frameScore;
+      this.currentFrame++;
+    }
```
Comment (z-rasool) [inline bowling.js:41]: There's a lot going on in this function, which is understandable as it the playing of the game. It might help to consider either refactoring the function into two, or moving some of the let statements to the top and including some empty lines to improve readability.

File: bowling.test.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  
+    it("does not score extra balls for frames 1-9 ", () => {
+      const game = new BowlingGame(10, 2, 10);
+      game.scores = [8, 0, 0];
+      game.getRandomScore = jest.fn(() => 10);
+      game.playExtraBalls();
+      expect(game.scores[0]).toBe(8);
+    });
+  });
+
```
Comment (z-rasool) [inline bowling.test.js:48]: This is a great coverage. Perhaps you could also consider some edge cases e.g. what happens if one of the scores entered isn't a number?

**Review by z-rasool — COMMENTED** [review]
I really like your solution, you've managed to use just one class and include most of the functionality in the requirements without overcomplicating things. Well done.

As an extension, you could look at how you could separate this into 2 or 3 classes.
